pravega/pravega
# Issue 4107: Automatic delegation token renewal

**Change log description**  
Introduces automatic token renewal. 

**Purpose of the change**  
Resolves #4107. 

**What the code does**  
Here's what the code does, at a high level:
1. Adds a `DelegationTokenProvider` (as well as related factory and implementation classes) that is used as a proxy for a delegation token in places where a string delegation token was used. 

   For JWT tokens, when the token is retrieved via the `retrieveToken()` method, the provider checks if the delegation token is nearing expiry. If the token is nearing expiry, it refreshes the token, i.e., fetches a new token and returns that one. Otherwise, it returns an existing token. 
   
2. Adds a token handling strategy for tackling these different scenarios: 
    * A JWT token
    * An empty token which occurs when auth is disabled. This strategy is also used in many test cases. 
    * A non-JWT token, which is used in some of the tests. 
    * A null token for cases where a delegation token is not populated in advance and is fetched lazily when needed. 

**How to verify it**  
All tests should pass.

I used the following temporary test suite to verify the changes: 

```java
package io.pravega.test.integration;

import io.pravega.client.security.auth.DelegationTokenProviderFactoryTest;
import io.pravega.client.security.auth.JwtBodyTest;
import io.pravega.client.security.auth.JwtTokenProviderImplTest;
import io.pravega.client.segment.impl.AsyncSegmentInputStreamTest;
import io.pravega.client.segment.impl.ConditionalOutputStreamTest;
import io.pravega.client.segment.impl.SegmentMetadataClientTest;
import io.pravega.client.state.impl.RevisionedStreamClientTest;
import io.pravega.client.stream.impl.EventStreamReaderTest;
import io.pravega.client.stream.impl.SegmentSelectorTest;
import io.pravega.segmentstore.server.host.handler.PravegaRequestProcessorAuthFailedTest;
import io.pravega.shared.protocol.netty.WireCommandsTest;
import org.junit.runner.RunWith;
import org.junit.runners.Suite;

@RunWith(Suite.class)
@Suite.SuiteClasses({
        // New unit test classes focused on the enhancement
        JwtTokenProviderImplTest.class,
        DelegationTokenProviderFactoryTest.class,
        JwtBodyTest.class,

        // Some of the relevant existing unit test classes
        EventStreamReaderTest.class,
        RevisionedStreamClientTest.class,
        AppendTest.class,
        AppendReconnectTest.class,
        ReadTest.class,
        SegmentSelectorTest.class,
        ConditionalOutputStreamTest.class,
        PravegaRequestProcessorAuthFailedTest.class,
        WireCommandsTest.class,
        AsyncSegmentInputStreamTest.class,
        SegmentMetadataClientTest.class,

        // New and existing integration tests related to delegation tokens
        DelegationTokenTest.class
})

public class DelegationTokenTestSuite {
}
```
